Add variant for events that have been redacted

DevinR528 opened the pull request:
This adds a new content type `RedactedContent` that contains the reason field, the redacted events event type string, and a `BTreeMap<String, JsonValue>` for the few events that keep content keys. The `Any*Event` enums will return `Self::IsRedacted(Event)` early when an event with a redacted_because field is encountered.

sphinxc0re commented:
May, I say, that I'm impressed by your solution 😄
Just a few days ago I thought of putting the chain implementation into a macro but couldn't find an obvious way to do it.
Also, if it matters in any way, from my side the code looks pretty good. This might be good for a merge.

sphinxc0re commented on src/api/r0/chain.rs:
```diff
@@ -0,0 +1,25 @@
+/// `chain_impl!(JoinRoom, []);`
+#[macro_export]
+macro_rules! chain_impl {
+    ($chain:ident, [ $($middleware:ident),+ ]) => {
```
Okay, one thing: Why is here a `+` marking that this section contains `n` entries whereas `n > 0`. Does that mean, there can't be an implementation without a prechained middleware? What about the `/publicRooms` endpoint? It is completely public.

jplatte replied:
Why would you have a Chain at all, then?

sphinxc0re replied:
Good point 😄

mujx replied:
You would always need to implement `chain`. Look at `src/server.rs` how the endpoints are registered.

sphinxc0re replied:
Ah, there was my point 😂

farodin91 replied:
I'll extend this for empty chain's

jplatte replied:
@mujx Ah okay, so that would be for being able to work in a uniform way with both chained and non-chained endpoints? Makes sense.

jimmycuadra commented:
This looks good to me. Ready for it to be merged?

farodin91 commented:
One small thing extend the macro, for no middleware.

mujx commented:
Maybe it would be better to move this into `src/macros.rs` instead. Following the conventions.

sphinxc0re commented:
@mujx no I think `middleware!` is too arbitrary and `chain!` doesn't really tell what the macro does. I like the current naming.

farodin91 commented:
@jimmycuadra If you like it, you could merge it.

jimmycuadra commented:
If we're going to use a macro to create these impls, is there any benefit to adding this `RequestChain` trait? Seems to just add another import to every API module.

farodin91 commented:
In docs you could look for all structs that implement `chain`.

jimmycuadra commented on src/api/r0/account.rs:
```diff
@@ -126,19 +109,7 @@ impl Handler for DeactivateAccount {
-        chain
-    }
-}
+chain_impl!(PutAccountData, [JsonRequest, UserIdParam, DataTypeParam, AccessTokenAuth]);
```
Missing `RoomIdParam` here?

jimmycuadra commented on src/api/r0/chain.rs:
```diff
@@ -0,0 +1,26 @@
+
```
Seconded. Kill the extra lines.

jimmycuadra commented on src/api/r0/chain.rs:
```diff
@@ -0,0 +1,26 @@
+}
+
+/// RequestChain
+pub trait RequestChain {
```
Let's rename this trait to `MiddlewareChain`, the macro to `middleware_chain!`, and move this whole file to src/middleware/chain.rs.

jimmycuadra commented on src/api/r0/login.rs:
```diff
@@ -19,9 +20,9 @@ struct LoginResponse {
 pub struct Login;
 
-impl Login {
+impl RequestChain for Login {
```
Any reason this one doesn't use the macro?

mujx commented:
There are some consistency issues. In some files, like `src/api/r0/profile.rs`, the macro is replacing the existing chain function right before the handler implementation while in some other files, like `src/api/r0/account.rs`, the macro was moved up to the struct definition.

A universal format for every endpoint could be:
1.`structs`
2.`middleware_chain!`
3.`handler`

sphinxc0re commented:
@mujx sounds reasonable. Would it be a good idea to put this order into the contribution guide?

farodin91 commented:
@jimmycuadra Resolved all issues.

DevinR528 committed:
Add variant for events that have been redacted

This adds a new content type RedactedContent that contains the reason
field and holds the redacted events event type string. The Any*Event
enums will return Self::IsRedacted(Event) early when an event with a
redacted_because field is encountered.

DevinR528 commented:
cc @poljar @jplatte 

This would work for matrix-sdk, it keeps the ability to swap out content, serializes and deserializes without losing the original event type string, and leaves the original event keys untouched.

may resolve #36

DevinR528 committed:
Move redacted tests into their own file

DevinR528 committed:
Fix clippy cmp-owned warnings remove duplicate redaction test

poljar commented:
This is different from the proposed solution in the issue, no? Could you explain the benefits of it? Seems like we're throwing away  typing quite a bit compared to the proposed solution, not that it's super important since you're gonna handle most redacted event types the same way.

DevinR528 commented:
> Seems like we're throwing away string typing quite a bit compared to the proposed solution,

What is string typing? Is this referring to returning early when deserializing? A better way to handle [this](https://github.com/ruma/ruma/pull/112/commits/5bf00c84c80e81f8a9fdc4748a0f9cfc24cf3ee4#diff-391bc0ba5e259f9d1e824a8fdce67f37R221) would be to check the event type string then check if it contains `redacted_becasue` so inside of each match arm there would be an `if else` block checking if it was a redacted event else deserializing as normal.

Would that address the string typing?

poljar commented:
> What is string typing? Is this referring to returning early when deserializing? A better way to handle [this](https://github.com/ruma/ruma/pull/112/commits/5bf00c84c80e81f8a9fdc4748a0f9cfc24cf3ee4#diff-391bc0ba5e259f9d1e824a8fdce67f37R221) would be to check the event type string then check if it contains `redacted_becasue` so inside of each match arm there would be an `if else` block checking if it was a redacted event else deserializing as normal.

Err, the string is a leftover that I forgot to remove. I meant we're producing the same type for every redacted event type, while the issue suggested to have a couple of types for the events for the events that need special handling.

That method to check if the event is redacted sounds good, but I'm wondering if we want to produce just one redacted type or a `RedactedMemberEvent`, `RedactedPowerLevelEvent`, etc... and a generic one for the events that aren't specially handled during redaction.

jplatte commented:
Yeah, I'm pretty sure this needs further discussion before I can merge any implementation.

DevinR528 commented:
#114 will implement redacted events
